Android.mk

```diff
@@ -140,6 +140,155 @@ $(LOCAL_BUILT_MODULE): FORCE
+				for target in $$dupes_target; do \
+					echo "$${vendor_src}:$${target}" >> $@; \
+				done; \
+			fi; \
+		fi; \
+	done;
+
+# Parameters for rebuilding filesystem
+# We hardcode sparse for now
+.PHONY: $(TARGET_CMDLINE_FS)
+$(TARGET_CMDLINE_FS):
+	echo "system.img: -s -l $(BOARD_SYSTEMIMAGE_PARTITION_SIZE) -a system" > $@
+
+# Building the list of blobs we want to delete from the ZIP file
+# So we take the target files from the blob map and we mangle them to match
+# what will be done in the ZIP file
+.PHONY: $(TARGET_BLOBS_DELETE_LIST)
+$(TARGET_BLOBS_DELETE_LIST): $(TARGET_LISTFILES_BLOBS) $(TARGET_BLOBS_MAP)
+	@rm $@ || touch $@; \
+	for l in $$(cat $(TARGET_BLOBS_MAP) | cut -d':' -f2 | sed -e 's|^root/|ramdisk/|g'); do \
+		for b in $$(grep -i "$$l$$" $(TARGET_LISTFILES_BLOBS)); do \
+			echo $$b >> $@; \
+		done; \
+	done; \
+	sort -o $@ < $@; mv $@ $@.tmp; uniq < $@.tmp > $@; rm $@.tmp
+
+# Building the list of blobs we will want to reinject, from the device, so we
+# need to make use of the source of the blob map
+# We need to retransform ramdisk/ into /
+.PHONY: $(TARGET_BLOBS_INJECT_LIST)
+$(TARGET_BLOBS_INJECT_LIST): $(TARGET_LISTFILES_BLOBS) $(TARGET_BLOBS_MAP)
+	rm $@ || touch $@; \
+	for map in $$(cat $(TARGET_BLOBS_MAP)); do \
+		blob_in_mk=$$(echo "$$map" | cut -d':' -f1); \
+		blob_in_fs=$$(echo "$$map" | cut -d':' -f2); \
+		source=$$(grep -m1 "^$$blob_in_mk:" "$(TARGET_DEVICE_BLOBS)" | cut -d' ' -f1 | cut -d':' -f2); \
+		l=$$(echo "$$blob_in_fs" | sed -e 's|^root/|ramdisk/|g'); \
+		for b in $$(grep -i "$$l$$" $(TARGET_LISTFILES_BLOBS)); do \
+			reall="/$$(echo $$source | sed -e 's|^root/||g')"; \
+			echo "$$reall:$$b" >> $@; \
+		done; \
+	done; \
+	sort -t ':' -k 2 -o $@ < $@; mv $@ $@.tmp; uniq < $@.tmp > $@; rm $@.tmp
+
+# This will produce the blobfree main distribution file. We rely on the make
+# target "target-files-package" that produces intermediate ready-to-use ZIP
+# file and tree from which we can build upon.
+$(TARGET_BLOBFREE_ZIP): target-files-package $(TARGET_BLOBS_DELETE_LIST)
+	@cp $(BUILT_TARGET_FILES_PACKAGE) $@ && \
+	zip --delete $@ $$(tr '\n' ' ' < $(TARGET_BLOBS_DELETE_LIST)); \
+	zip --delete $@ "OTA/*" "DATA/*" "RADIO/*" "META/*"
+
+.PHONY: $(TARGET_LISTFILES_BLOBS) $(TARGET_LISTFILES_NOBLOBS) compare-zipfiles
+$(TARGET_LISTFILES_BLOBS): target-files-package
+	@zipinfo -l2 $(BUILT_TARGET_FILES_PACKAGE) > $@
+
+$(TARGET_LISTFILES_NOBLOBS): $(TARGET_BLOBFREE_ZIP)
+	@zipinfo -l2 $(TARGET_BLOBFREE_ZIP) > $@
+
+compare-zipfiles: $(TARGET_LISTFILES_BLOBS) $(TARGET_LISTFILES_NOBLOBS)
+	diff -uw $(TARGET_LISTFILES_BLOBS) $(TARGET_LISTFILES_NOBLOBS) || true
+
+# Target controlling the build of the {user,addon}-facing ZIP distribution file,
+# which will just contain all the dependencies within the root of the archive.
+$(TARGET_BLOBFREE_PKG): $(TARGET_BLOBFREE_ZIP) $(TARGET_BLOBS_INJECT_LIST) $(TARGET_CMDLINE_FS) $(INSTALLED_DTIMAGE_TARGET) $(TARGET_DEVICES_JSON) $(recovery_fstab)
```

Comment:
There might be some kind of race condition here, with recovery_fstab. It's being set by build/core/Makefile but I have noticed that sometimes it is not. TARGET_RECOVERY_FSTAB is valid in all cases, though. I think we do define it on all the ports we have, so ...